File: implement-shell-tools/cat/cat.py
import argparse

parser = argparse.ArgumentParser(
    prog="cat",
    description="read, display, and concatenate text files.",
)

parser.add_argument("-n", action="store_true", help="Number all output lines.")
parser.add_argument("-b", action="store_true", help="Number non-blank output lines.")
parser.add_argument("paths",nargs="+", help="The file to search")

args = parser.parse_args()

for path in args.paths:
    try:
        with open(path, mode='r', encoding='utf-8') as f:
            lines = f.readlines()

    except Exception as err:
        print(f"Error reading file '{path}': {err}")
        continue

    line_num = 1

    for line in lines:
        if args.b:
            if line.strip() != "":
                print(f"{line_num:5} {line}", end="")
                line_num += 1
            else:
                print()
        elif args.n:
            print(f"{line_num:5} {line}", end="")
            line_num += 1
        else:
            print(line, end="")

File: implement-shell-tools/cat/cat.js
import { promises as fs } from "node:fs";
import { program } from "commander";

program
  .name("myCat")
  .description("The re-implementation of the cat command")
  .argument("<files...>", "files to read")
  .option("-n, --number", "number all lines")
  .option("-b, --number-nonblank", "number non-empty lines");

program.parse();

const options = program.opts();
const paths = program.args;
let lineNumber = 1;

try {
    for (const path of paths) {
        let content = await fs.readFile(path, "utf-8");
        const lines = content.split("\n");
        if (options.numberNonblank) {
            content = lines.map(line => {
                if (line.trim() === "") return line;
                return `${lineNumber++} ${line}`;
            }).join("\n");
        } else if (options.number) {
            content = lines.map((line, index) => `${index + 1} ${line}`).join("\n");
        }
    
        console.log(content);
    }
} catch (err) {
    console.error(`Error reading file: ${err.message}`);
    process.exit(1);
}